### DISCUSSION TITLE
Neolink 0.5.11 New parameters

### DISCUSSION THREAD

**Discussion by Veuchez** [QuantumEntangledAndy/neolink#84] [Q&A]
How do the new configuration parameters introduced with version 0.5.11 work?
`buffer_size = 100`
`use_smoothing = true`
What values can be put? Is there a minimum or maximum range?
Thank you!

**Comment by QuantumEntangledAndy** [comment]
Ah smoothing is true/false and I think I put a min max range of 10-500 on the buffer_size

**Reply by QuantumEntangledAndy** [reply]
If I get the time. I'm going to make a no buffer mode too. Where it will just send the frames on as they are received disregarding the time stamps.

**Comment by Veuchez** [comment]
Is 10-500 a value without a unit of measurement? 
I do not understand how to calibrate it. 
With "100" and 2 cameras connected, initially it's fine, after a few hours I buffer exhausted.

**Reply by QuantumEntangledAndy** [reply]
They represent frames. They should depend upon the fps of the camera.

If you see exhaustion after hours then changing it won't help. From my tests under normal operation the buffer increases/decreases in size by only about +-5 frames. So over the hours it was probably mostly constant.

If it goes -5 then almost instantly (less than a second) it can refill back to +5.

**Reply by QuantumEntangledAndy** [reply]
Is it by any chance after 71 minutes?

**Reply by Veuchez** [reply]
I haven't looked after exactly how much, I'll look at the next log.
Is there a specific reason after 71 minutes?

**Reply by QuantumEntangledAndy** [reply]
Yeah the camera uses 32bit integers for its times stamps which is in microseconds

The maximum u32 is 4294967295. 

4294967295us = 4294967295/1000ms = 4294967295/1000/1000s = 4294967295/1000/1000/60 mins ~= 71 mins

**Reply by QuantumEntangledAndy** [reply]
Most apps use u64 for microseconds to avoid this but it seems that the cameras are limited to u32

**Reply by QuantumEntangledAndy** [reply]
There should already be code in there to deal with this situation but if it is 71 mins then it gives me a hint of where to look for issues

**Reply by Veuchez** [reply]
Hi, sorry, but I still haven't understood the 'buffer_size' parameter properly. 
It ranges from 0 to 500, but what is the default value? 
If I set it to zero, everything becomes jerky.

**Reply by QuantumEntangledAndy** [reply]
It's a buffer of frames designed to smooth over network issues between the camera and neolink.

The value you need depends on how much frames might be getting lost in transit. A good connection should only need a few frames. 

The live time is set a few seconds behind and frames are kept stored (in the buffer) until their live time comes up. This will smooth jerkiness as it will release the frames at exactly the right moment. 

Having it on zero completely disables the buffer and frames are sent through to the client immediately upon receipt. You get the frames as neolink gets them but you also have to deal with the jerkiness

**Reply by Veuchez** [reply]
Thank you very much! 
Is there a way to calculate the correct value, or do I have to do it through trial and error? 
Does the value I input refer to each individual camera that I set it for? 
Just to clarify, my main stream is set at 15fps, and the substream is also set at 15fps. 
What criteria could I follow?

**Reply by QuantumEntangledAndy** [reply]
Trial and error for all cameras but a too large value won't cause issues except pushing the live time backwards more.

**Comment by Veuchez** [comment]
Sorry for the delay. 
Here is a log with 2 cameras active at 2048kbps. 
[log2048kbps.txt](https://github.com/QuantumEntangledAndy/neolink/files/11545091/log2048kbps.txt)
I have several interruptions, initially very distant from each other, then after about 70 minutes the interruptions become more and more frequent. 
Maybe that's why you said?
I also attach the log with 2 cameras at 1536kbps. 
[log1536kbps_.txt](https://github.com/QuantumEntangledAndy/neolink/files/11545096/log1536kbps_.txt)
After an hour at 1536kbps I have to restart neolink because it becomes unusable. 
Restarting it everything goes back to fluid.
These tests are done by watching streaming on VLC. On Frigate I think I'm doing something wrong because I see well, but if I activate the audio everything becomes a mess ...

**Reply by Veuchez** [reply]
@QuantumEntangledAndy Are you seeing anything strange in my logs by any chance?

**Reply by QuantumEntangledAndy** [reply]
I'm kinda on holiday at the moment. I do some easier things in between holiday things. But looking for possible unknown issues with unknown solutions is not something I have the time to tackle right now.
